feat(content-group-banner): add Content Group Banner component

proeung opened the pull request:
### Related Ticket(s)

DDS Tickets: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5354 & https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5506
HC JIRA Ticket: https://jsw.ibm.com/browse/HC-1736

### Description
- This PR adds the "Content Group - Banner" layout component/pattern that's needed in the new Solution page template that will be implemented in the AEM CMS.
- This component is a variant of HC's [Module 12A](https://pages.github.ibm.com/MSC-Cloud/hc_fractal/components/detail/12a--default.html) and this iteration is the first version that will be going into the MVE state of the Solution template.

### Design Spec
https://ibm.box.com/s/tqp7c0wm99ha0b7jrm0sqhq3whwdjv84
![Screen Shot 2021-04-01 at 11 10 28 AM](https://user-images.githubusercontent.com/1815714/113315076-e6583b80-92da-11eb-92bf-a23923a57953.png)


### Changelog

**New Components**

- `<dds-content-group-banner>` – a banner-like pattern that is a variant of `<dds-content-group>`, and includes a heading and link list.
- `<dds-content-section-banner>` – a Content Group Banner component assembled in a section with a left-column header.

**Removed**

- N/A

**Testing Instructions**
- Visit the `Content Group Banner` - `/?path=/story/components-content-group-banner--default`
- Ensure that the component matches with the design specs for all breakpoints.
- Visit the `Content Section Banner` - `/?path=/story/components-content-section-banner--default`
- Ensure that the card section headline renders correct and matches with the design specs for all breakpoints.

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "package: styles": Carbon Expressive -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

shixiedesign commented:
Hi @proeung! Thank you! This looks great. Only a few minor comments: 

1. This component should not allow "Jump link" as a CTA type
![image](https://user-images.githubusercontent.com/15144993/113452547-35cd6300-93d2-11eb-8881-6118f7596287.png)
Here's func spec: 
![image](https://user-images.githubusercontent.com/15144993/113454365-10daef00-93d6-11eb-9696-0f7552984597.png)

2. We shouldn't have the option for "None" as a CTA type. This might be a bug from Link list. If out of scope please lemme know, I will make new issue to fix. 
![image](https://user-images.githubusercontent.com/15144993/113453010-2864a880-93d3-11eb-835f-428d15de8730.png)

3. Grid alignment of link list is wrong. Vertical rule should be in the gutter, and text label starting on grid. 
Current
![image](https://user-images.githubusercontent.com/15144993/113453074-46caa400-93d3-11eb-98f7-89da95b3a4d8.png)
Expecting
![image](https://user-images.githubusercontent.com/15144993/113453141-682b9000-93d3-11eb-9ac6-69d9ea30d04a.png)

4. Focus state seems funky. Should be simple 2px solid inset outline in `$focus`, and only width of the link
![image](https://user-images.githubusercontent.com/15144993/113452610-5a293f80-93d2-11eb-9c5d-639924b548ee.png)
Sometimes focus becomes orange... 
![image](https://user-images.githubusercontent.com/15144993/113453723-9eb5da80-93d4-11eb-9304-021112d0b5f1.png)

5. Grid at `md` breakpoint, content should only occupy 6 out of 8 col 
**Current**
![image](https://user-images.githubusercontent.com/15144993/113453607-572f4e80-93d4-11eb-81d7-738d9a6b353b.png)
**Expecting**
![image](https://user-images.githubusercontent.com/15144993/113453624-61514d00-93d4-11eb-9d93-61f63b522b4d.png)

6. Last knitpick, please use sentence case for the story name: "Content group banner" 
![image](https://user-images.githubusercontent.com/15144993/113454294-e7ba5e80-93d5-11eb-93f8-84094facb167.png)


Thank you!

proeung commented:
@shixiedesign Thanks for doing the design QA for this PR! I went in and addressed all of the feedback that you've listed above, except for the focus state styling issue. I'm not able to replicate this issue and this particular PR is just using the default focus state style that has been added to the `link-list` component. Feels like this might be something we'll need to address there so that it triggers down to this component.

Also, I added a component for the variant where this content group banner is being used in the `Card Section` with left sidebar headline. Can you take a look and let me know what you think of this variant as well? Thanks! https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/5697/index.html?path=/story/components-card-section-banner--default

shixiedesign commented:
Hey Putra, the Content group banner is in excellent shape! I will open a ticket to fix link list focus state. 

Re: card section banner though... 😅 , it shouldn't be called "Card section banner", and more likely should be "Content section banner". Card section would require the section to contain some type of card. I'd like @photodow to weigh in on this. If we add it as a component, we'd have to prepare a bit of visual and functional spec (or catch up on making sure the design spec side don't go out of sync with code). I know this is what AEM needed. I just thought it would be assembled on AEM side. So hang on a sec while I discuss with design team. 

(If you separate out Content section banner as a new PR, we can merge this one for Content group banner!)

photodow commented:
> If we add it as a component, we'd have to prepare a bit of visual and functional spec (to making sure the design spec side don't go out of sync with code).  ... I just thought it would be assembled on AEM side. 

Agreed. AEM should be able to achieve what they need using content group banner and content section. I don't think we should / can support every section variation. But if that's wrong we should address in a feature request because like @shixiedesign mentioned there is a lot of documentation that will need to be created to support it.

proeung commented:
> Re: card section banner though... 😅 , it shouldn't be called "Card section banner", and more likely should be "Content section banner". Card section would require the section to contain some type of card.

@shixiedesign That's a good point. I originally had this component as "Card section banner" because I want to keep the naming convention the same as the other components where the left sidebar headline is being used, however, I realized that having the prefix word `card` doesn't make sense.

![Screen Shot 2021-04-06 at 2 48 58 PM](https://user-images.githubusercontent.com/1815714/113763409-aa144900-96e7-11eb-913e-43bcabb9ae35.png)

I just went ahead and rename the component to "Content section banner" so that it's more consistent with the design and naming convention that you guys have established (see commit - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5697/commits/ea4c8570fd83fbff1ef0e27a4d703e92c453563c).

![Screen Shot 2021-04-06 at 2 49 12 PM](https://user-images.githubusercontent.com/1815714/113765383-30318f00-96ea-11eb-8e3d-9304439f5148.png)
**Preview Link:** https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/5697/index.html?path=/story/components-content-section-banner--default


@shixiedesign @photodow As for your question on whether or not to have the `<dds-content-section-banner>` component be available within the library, I would like to note that simply taking the custom elements of `<dds-content-group-banner>` and `<dds-content-section-banner>` to assemble a variant where there's a section headline requires extra styling dependency. In this PR, I've included a style partial that accounts for the different padding and spaces that are needed for this component to align with the design specs (see [_content-section-banner.scss](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5697/files#diff-65dc4d89dac1f3d98bffe47992c39b615e274992568267f1102d190d4a010b11R8)). Also, I'd like to create little work as much as possible on the AEM side so that we can eliminate one-off styling and overriding.

In addition, I anticipate that we'll eventually add other attributes/props to this `<dds-content-section-banner>` component that will allow for full-width background image as well as gradient options similar to [Module 12A](https://pages.github.ibm.com/MSC-Cloud/hc_fractal/components/detail/12a--default.html) that we have within the HC Fractal Library. Regardless, I'm happy to take out `<dds-content-section-banner>`, if the design team feels that this introduces new scope of work that hasn't been accounted for or required.

Also, I'd like to get a code review for this PR. @jeffchew @andysherman2121 @annawen1 Can you help take a look? Thanks!

photodow commented:
@proeung can you elaborate on `requires extra styling dependency.`?

jeffchew commented on packages/web-components/src/components/content-section-banner/__stories__/README.stories.mdx:
````diff
@@ -0,0 +1,63 @@
+### HTML
+
+```html
+<dds-content-section-banner heading="Aliquam condimentum interdum">
````
@proeung I'm wondering if this could possibly be either a slot or element instead of being passed as an attribute, since we'd want to have lightDOM be available from an SEO perspective for non Googlebot crawlers.

proeung replied:
@jeffchew  The `<dds-content-section-heading>` is already defined as a slot. I just pushed up a commit that alters the README file of the `<dds-content-section-banner>` component so that folks understand that they can use this as a slot or as the light DOM with heading tags. See here - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5697/commits/e8943a4e71d8b985ea9726989221fad63707600c.

ariellalgilmore commented on packages/web-components/src/components/content-group-banner/__tests__/content-group-banner.test.ts:
```diff
@@ -0,0 +1,61 @@
+      render(
+        template({
+          children: html`
+            <dds-content-group-banner>
+              <dds-content-group-heading>heading-foo</dds-content-group-heading>
+              <dds-link-list slot="footer" type="vertical">
+                <dds-link-list-item-cta icon-placement="right" href="https://www.ibm.com" cta-type="${CTA_TYPE.LOCAL}">
+                  cta-copy-foo
+                </dds-link-list-item-cta>
+                <dds-link-list-item-cta icon-placement="right" href="https://www.ibm.com" cta-type="${CTA_TYPE.EXTERNAL}">
+                  cta-copy-foo
+                </dds-link-list-item-cta>
+              </dds-link-list>
+            </dds-content-group-banner>
```
```suggestion
              <dds-content-group-heading>heading-foo</dds-content-group-heading>
              <dds-link-list slot="footer" type="vertical">
                <dds-link-list-item-cta icon-placement="right" href="https://www.ibm.com" cta-type="${CTA_TYPE.LOCAL}">
                  cta-copy-foo
                </dds-link-list-item-cta>
                <dds-link-list-item-cta icon-placement="right" href="https://www.ibm.com" cta-type="${CTA_TYPE.EXTERNAL}">
                  cta-copy-foo
                </dds-link-list-item-cta>
              </dds-link-list>
```

andysherman2121 replied:
@proeung Other than Ari's comment on the test file. The code looks good to me!

proeung replied:
@ariellalgilmore @andysherman2121 Great catch! https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5697/commits/83d1c2bb51c3574cb9177621089b97e7053d058b

proeung commented:
> @proeung can you elaborate on requires extra styling dependency.? Also, background props and such aren't in scope for this work, and might be better applied to our content section component. We'll have to take a look when that need arises.

@photodow The styling dependency is defined here [_content-section-banner.scss](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5697/files#diff-65dc4d89dac1f3d98bffe47992c39b615e274992568267f1102d190d4a010b11R8). In order to align with the [design specs](https://ibm.ent.box.com/file/789015932954) for the variant where there's a section headline (aka. Content Section Banner), we need to adjust the paddings of the content section component and ensure that it looks good across different breakpoints (see attached). This is my main reason for including the `<dds-content-section-banner>`, however, if that doesn't make sense. I'm happy to remove it from this PR.

![content-section-banner](https://user-images.githubusercontent.com/1815714/113921477-0fce0700-97b4-11eb-8bf5-954b9e18caf2.jpg)

photodow commented:
@proeung since it's not in scope for this work, can we pull it out and create an issue for it so we can work through these styling issues you're bringing up, and either decide from there if we need the section component or change something to make this work?

proeung commented:
@photodow Sounds good! I just pushed up a commit that removes all of the code related to the `<dds-content-section-banner>` (see - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5697/commits/56ec8428bade97c349dccb60baa9be4bd74d6ab9).

Depending on what's the solution is for this ticket (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5734), I'm hoping that we can at least repurpose the changes in this PR or use the finding here as references for the styling inconsistency seen within `<dds-content-section>`. Overall, I think PR just needs another design QA and we should be good for merging down.

shixiedesign commented:
Really appreciate it Putra! Awaiting the builds...

photodow reviewed (commented):
- [ ] Is this full width focus a link list problem?

![image](https://user-images.githubusercontent.com/3793636/114047120-0c3b8e00-984f-11eb-8966-ba3e1e520b90.png)

- [ ] Spacing on your small breaking points looks off. I think I see why you did it that way. The design specs here aren't clear, and I'd actually love for @shixiedesign to weigh in on this real quick. It looks like it goes down to the `16px`, but even that feels really tight for this. Almost like it was a mistake.

![image](https://user-images.githubusercontent.com/3793636/114047624-6f2d2500-984f-11eb-9132-1844b4df25ec.png)

- [ ] It also appears as if link list is bringing in some spacing with its last item that we need to consider here.

![image](https://user-images.githubusercontent.com/3793636/114047880-a6033b00-984f-11eb-877f-6a3634d9e92c.png)

- [ ] My guess is this last one is not apart of your PR, but I'm going to call it out here in context and we can probably open up a seperate issue to dive in.

For some reason the icons are changing size almost like at a percentage/fluid rate.

![Apr-08-2021 09-52-03](https://user-images.githubusercontent.com/3793636/114048512-42c5d880-9850-11eb-8426-fcc596036dc4.gif)

shixiedesign commented:
Quite a few of these are [`Link with icon` bug issue ](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/5705). 

But I'd second fixing: 

1. video CTA should still trigger media viewer lightbox etc. If this is too hard, let’s remove video as a CTA type for this PR
2. margin top and bottom of Content group banner should be:

| Breakpoint | `max` | `xlg` | `lg` | `md` | `sm` |
| --- | --- | --- | --- | --- | --- |
| Margin top | `spacing-07` 32px | `spacing-07` 32px | `spacing-07` 32px | `spacing-07` 32px | `spacing-07` 32px |
| Margin bottom | `spacing-09` 48px | `spacing-09` 48px | `spacing-09` 48px | `spacing-09` 48px | `spacing-07` 32px |

SO CLOSE. Thank you @proeung for the hard work!

proeung commented:
@shixiedesign I just added a commit to address the two minor changes you've noted above. Please take a look after the preview has finished building.

> 1. video CTA should still trigger media viewer lightbox etc. If this is too hard, let’s remove video as a CTA type for this PR

The "Video" CTA type should now open the lightbox media viewer.

> 2. margin top and bottom of Content group banner should be:

The spacing has been adjusted for all breakpoints.

shixiedesign reviewed (approved):
👍 Big thank you!!!

proeung commented:
@shixiedesign No prob! Thanks for the design review!

jeffchew commented on packages/web-components/src/components/content-group-banner/__stories__/content-group-banner.stories.ts:
```diff
@@ -0,0 +1,124 @@
+  title: 'Components/Content group banner',
+  decorators: [
+    story => html`
+      <dds-video-cta-container class="dds-ce-demo-devenv--simple-grid dds-ce-demo-devenv--simple-grid--content-group-banner">
```
A bit confused that the wrapper here is `dds-video-cta-container`...

proeung replied:
@jeffchew The `<dds-video-cta-container>` is needed in order for the video CTA type to open as an overlay. I'm just following the example shown here ( see - https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/master/packages/web-components/src/components/link-list/__stories__/link-list.stories.ts#L268). Also, I can wrap it around the content group banner component as well.

```
     <dds-video-cta-container>
          <dds-content-group-banner>
            <dds-content-group-heading>${heading}</dds-content-group-heading>
            
            <dds-link-list type="vertical" slot="complementary">
              <dds-link-list-item-cta
                icon-placement="${iconPlacement}"
                href="${ifNonNull(href)}"
                cta-type="${ifNonNull(ctaType)}"
                download="${ifNonNull(download)}"
              >
                Learn more about Kubernetes
              </dds-link-list-item-cta>
              <dds-link-list-item-cta
                icon-placement="${iconPlacement}"
                href="${ifNonNull(href)}"
                cta-type="${ifNonNull(ctaType)}"
                download="${ifNonNull(download)}"
              >
                Containerization A Complete Guide
              </dds-link-list-item-cta>
            </dds-link-list>
          </dds-content-group-banner>
        </dds-video-cta-container>
```

jeffchew replied:
That's fine, think we can leave as-is, thanks!

ariellalgilmore reviewed (commented):
<img width="930" alt="Screen Shot 2021-04-08 at 3 13 27 PM" src="https://user-images.githubusercontent.com/20210594/114102904-065ca180-987d-11eb-81fb-a4315e5c390d.png">
Hi! sorry i'm seeing this weird focus when I change the cta-type in knobs

proeung commented:
@ariellalgilmore @jeffchew @shixiedesign The extra focus state line issue that you're seeing when the `cta-type` knobs is changed is coming from the `link-list.scss` partial. I just set the `outline: none;` since it's not needed there. Let me know what you all think. https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/5697/commits/ed4649a1e46a91f823b1a994722030e06dbf356a#diff-ae714fb7b15c919dc94326e4024a99b9fbb7555737f90b636a1f72d046e1dc70R74

shixiedesign commented:
Yep made an issue to fix Link list #5746 
Out side scope of this PR 👍

jeffchew reviewed (approved):
LGTM!
